Add url creation to WidgetSettings

toger5 opened the pull request:
The Idea is that widgets usually are not created in the native code but by some external service (I think formerly this was called Dimensions) setting them in the state event. We are creating a virtual widget by not having a state event for element call but just creating the WidgetSettings.
The design I have in mind is that WidgetSettings represent the state event. And If you have the widget settings you can generate the Url you should use for the IFrame (This url can inject some additional data into the widget)

This is why I changed the WidgetSettings to be constructed using `new_virtual_widget`
in a normal non element call case this should look like this:

```rust
let settings = room.get_widgets()[0];
let url = settings.get_widget_url();
let w = Widget{
  settings: settings,
  ...
}
```
(I am considering making WidgetSettings its own module)


<!-- description of the changes in this PR -->

- [ ] Public API changes documented in changelogs (optional)

<!-- Sign-off, if not part of the commits -->
<!-- See CONTRIBUTING.md if you don't know what this is -->
Signed-off-by:

daniel-abramov commented on crates/matrix-sdk/src/widget/mod.rs:
```diff
@@ -50,6 +41,84 @@ pub struct Comm {
+    /// (`ContentLoad` message), or upon creation/attaching of the widget to
+    /// the SDK's state machine that drives the API.
+    pub init_on_load: bool,
+    /// This contains the url from the widget state event
+    /// In this url placeholders can be used to pass information from the client to the widget
+    /// Possible values are: `$widgetId`, `$parentUrl`, `$userId`, `$lang`, `$fontScale`, `$analyticsID`
+    /// 
+    /// # Examples
+    /// 
+    /// e.g `http://widget.domain?username=$userId`
+    /// will become: `http://widget.domain?username=@user_matrix_id:server.domain`
+    raw_url: Url,
```
Shouldn't it be something like [`Uri`](https://docs.rs/http/latest/http/uri/struct.Uri.html)?

toger5 replied:
what makes it a Uri? I think all the parts it can contain are conformant to a URL? Schme, Authority, Path, Query string, Fragment Identifier.

daniel-abramov commented on crates/matrix-sdk/src/widget/mod.rs:
```diff
@@ -50,6 +41,84 @@ pub struct Comm {
+    /// (`ContentLoad` message), or upon creation/attaching of the widget to
+    /// the SDK's state machine that drives the API.
+    pub init_on_load: bool,
+    /// This contains the url from the widget state event
+    /// In this url placeholders can be used to pass information from the client to the widget
+    /// Possible values are: `$widgetId`, `$parentUrl`, `$userId`, `$lang`, `$fontScale`, `$analyticsID`
```
OCD: Could you please add full stops at the end of the sentences? 🙂 Also I think it makes sense to run `cargo + nightly fmt`.

daniel-abramov commented on crates/matrix-sdk/src/widget/mod.rs:
```diff
@@ -50,6 +41,84 @@ pub struct Comm {
+    /// * `lang` - the language e.g. en-us
+    /// * `analytics_id` - This can be used in case the widget wants to connect to the 
+    /// same analytics provider as the client app only set this value on widgets which are known.
+    pub fn get_url(
+        &self,
+        room: JoinedRoom,        
+        parent_url: String,
+        font_scale: f64,
+        lang: String,
+        analytics_id: String,
+    ) -> String {
+        self.raw_url
+            .as_str()
+            .replace("$widgetId", &self.id)
+            .replace("$parentUrl", &encode(parent_url.as_str()))
+            .replace("$userId", &room.client().user_id().unwrap().to_string())
+            .replace("$lang", &lang)
+            .replace("$fontScale", &font_scale.to_string())
+            .replace("$analyticsID", &analytics_id)
+    }
```
I may be wrong about it (I think @jplatte may be more aware of how things are done in the Rust SDK and provide a more useful feedback), but typically I would expect that if `lang` is limited to some fixed inputs (`en-us`, `en-uk` etc), then `lang` should be an `enum`, the `parent_url` should probably be `http::Uri`.

Also, the `user_id().unwrap()` looks quite scary here (since there is no comment that says why it would never fail - it looks like it could panic).

toger5 replied:
Here I have the same question again. What exactly is the difference between a rust Uri and Url

toger5 commented:
Are the test sth I can fix?

Hywan commented on crates/matrix-sdk/src/widget/mod.rs:
```diff
@@ -50,6 +42,89 @@ pub struct Comm {
+    /// e.g `http://widget.domain?username=$userId`
+    /// will become: `http://widget.domain?username=@user_matrix_id:server.domain`.
+    raw_url: Url,
+}
+impl WidgetSettings {
```
```suggestion
}

impl WidgetSettings {
```

Hywan commented on crates/matrix-sdk/src/widget/mod.rs:
```diff
@@ -50,6 +42,89 @@ pub struct Comm {
+#[derive(Debug)]
+pub struct WidgetSettings {
+    /// Widget's unique identifier.
+    pub id: String,
+    /// Whether or not the widget should be initialized on load message
```
```suggestion
    pub id: String,

    /// Whether or not the widget should be initialized on load message
```

Hywan commented on crates/matrix-sdk/src/widget/mod.rs:
```diff
@@ -50,6 +42,89 @@ pub struct Comm {
+    /// Whether or not the widget should be initialized on load message
+    /// (`ContentLoad` message), or upon creation/attaching of the widget to
+    /// the SDK's state machine that drives the API.
+    pub init_on_load: bool,
+    /// This contains the url from the widget state event.
```
```suggestion
    pub init_on_load: bool,

    /// This contains the url from the widget state event.
```

Hywan commented on crates/matrix-sdk/src/widget/mod.rs:
```diff
@@ -50,6 +42,89 @@ pub struct Comm {
+            .replace("$analyticsID", analytics_id)
+    }
+
+    /// WidgetSettings are usually created from a state event.
```
```suggestion
    /// `WidgetSettings` are usually created from a state event.
```

Hywan commented on crates/matrix-sdk/src/widget/mod.rs:
```diff
@@ -50,6 +42,89 @@ pub struct Comm {
+    /// (currently unimplemented)
+    /// But in some cases the client wants to create custom `WidgetSettings`
+    /// for specific rooms based on other conditions.
+    /// This function returns a WidgetSettings object which can be used
```
```suggestion
    /// This function returns a `WidgetSettings` object which can be used
```

Hywan commented on crates/matrix-sdk/src/widget/mod.rs:
```diff
@@ -50,6 +42,89 @@ pub struct Comm {
+        init_on_load: bool,
+        base_path: &str,
+    ) -> Result<Self, url::ParseError> {
+        let raw_url = format!("{base_path}?widgetId=$widgetId&parentUrl=$parentUrl#?&userId=$userId&lang=$lang&fontScale=$fontScale&analyticsID=$analyticsID", base_path= base_path);
```
```suggestion
        let raw_url = format!("{base_path}?widgetId=$widgetId&parentUrl=$parentUrl#?&userId=$userId&lang=$lang&fontScale=$fontScale&analyticsID=$analyticsID");
```

If the capturing name has the same name as a variable, it automatically capture the variable.

Hywan reviewed (dismissed):
Small changes required, but overall, it looks good, thanks!

toger5 commented:
Is it an option to merge this without FFI for the new code?

Hywan commented:
It's fine for me to merge without the FFI bindings, as long as they continue to compile just fine. This is ensured by the CI anyway.

toger5 commented:
This still needs testing (a unit test should be very doable. Maybe @robintown can help confirming if all required features are present and the format is correct.

daniel-abramov reviewed (commented):
Hey @toger5,

If I get it right, this PR introduces 2 new concepts:
1. A function that transforms a "fully-qualified widget URL" (if I can call it like this) based on the existing valid `Url`, amended with some additional data. I.e. it's a function `fn (&Url, ...set of fields...) -> NewUrl`. Currently, it corresponds to the `WidgetSettings::get_url()`. I would write it differently though (see below).
2. A "constructor" function for the `WidgetSettings` (a structure that already exists in the `widget-api` branch, so I believe it was re-introduced here by mistake). That corresponds to the current `new_virtual_element_call_widget()` (I would do it differently too 🙂).

As for the topic (1), it looks to me (based on the current implementation) as if we want to have some sort of a `WidgetUrlBuilder` type that could be used like this:

```rust
let widget_url = WidgetUrlBuilder::new(url)
    .with_widget_id(&id)
    .with_matrix_info(&info) // device_id, user id, room id
    .with_ui_settings(&ui_settings) // font scale, language etc
    .build()?;
```

I feel like it would be easier to use and also the implementation will be a bit more elegant (i.e. it's akin to the [`ClientBuilder` from `hyper`](https://docs.rs/hyper/latest/hyper/client/struct.Builder) in some sense). My current problem with the `get_url` is that it accepts `WidgetSettings`, but actually it only needs the widget ID and the set of parameters it accepts is very large (and not strictly typed), also it e.g. accepts a `Room` object just for the sake of using a user ID from it. Moreover, it gets `room_id` and `Room` as 2 separate parameters that I find a bit odd in a sense that `Room` already contains an `id` so having separate `Room` and `id` parameters means that we can easily use one `Room` object, but a completely different disjoint room `id` with it.

As for topic (2), I need a bit more context to understand why/how it is supposed to work. It feels like a fancy constructor for the `WidgetSettings` and has a large set of parameters as well. I feel like it may be constructed similarly to (1), e.g. I see that it does most of the things that have already been done in (1), so perhaps it would be better to re-use (1) here and just amend it with a couple of missing fields instead of re-introducing the same logic again.

What do you think? Does it make any sense?

toger5 commented:
Most of the reasoning of the design is based on the fact, that widgets in general are constructed by a widget state event. So the widget settings are basically a representation of this state event.
For the state event a (weakly typed) url is provided to transport Information to the widget: userId, roomId (there is no other way for the widget to to figure out the room id)...
Therefoer a builder pattern does not make sense to me. The format of the Url is very generic It is possible to do things like: `my.site.com/$lang` it is not required to use url url parameters (`?myVal=$lang`). Hence there is no generic way to build such a url. The process is always "get something from the widget state event without any knowledge about the format" -> "replace all the keywords (`$`)" -> "open an IFrame with the url".
In element call we have the special case, that we know the url format. So for element call a Builder would make sense.

> A "constructor" function for the WidgetSettings (a structure that already exists in the widget-api branch, so I believe it was re-introduced here by mistake). That corresponds to the current new_virtual_element_call_widget() (I would do it differently too 🙂).

I cannot find the constructor in the widget-api branch. The structure already exists, I am only moving it out into a seperate file.
The long name `new_virtual_element_call_widget` was choosen to make it explicit that this is an edge case usage and that it is not normal to create WidgetSettings like this. Instead one just uses the widget state event.

> As for topic (2), I need a bit more context to understand why/how it is supposed to work. It feels like a fancy constructor for the WidgetSettings and has a large set of parameters as well. I feel like it may be constructed similarly to (1), e.g. I see that it does most of the things that have already been done in (1), so perhaps it would be better to re-use (1) here and just amend it with a couple of missing fields instead of re-introducing the same logic again.

This part is just faking the value of the state event. For a normal widget all this would be defined by the url provided in the state event. Here we just ask for the values seperately so we can control the format of the url on both platforms.

To sum up: This is mostly guaranteeing that the ffi api is mostly designed with normal widgets in mind that the rust sdk will generate from the room state. And just provide a very small change for the custom elemetn call virutal widget situation. Only the function `new_virtual_element_call_widget` is element call specific and everything else is shared and fits the architecture of "normal" widgets.
Otherwise we would bild a longer parallel/seperate logic chain for normal and virtual widgets which I tried to avoid.

toger5 commented on crates/matrix-sdk/src/widget/widget_settings.rs:
```diff
@@ -0,0 +1,126 @@
+    ///   to the
+    /// same analytics provider as the client app only set this value on widgets
+    /// which are known.
+    pub fn get_url(
+        &self,
+        room: Room,
+        parent_url: &str,
+        font_scale: f64,
+        lang: LanguageTag,
+        room_id: &str,
+    ) -> String {
```
(discussed with @daniel-abramov)
This should become a struct with widget parameters.

toger5 commented on crates/matrix-sdk/src/widget/widget_settings.rs:
```diff
@@ -0,0 +1,126 @@
+    /// * `preload` if set, the lobby will be skipped and the widget will join the call on the `io.element.join` action.
+    /// * `base_url` the url of the matrix homserver in use e.g. https://matrix-client.matrix.org.
+    /// * `analytics_id` can be used to pass a posthog id.
+    pub fn new_virtual_element_call_widget(
+        base_path: &str,
+        id: String,
+        embed: bool,
+        hide_header: bool,
+        device_id: &str,
+        preload: bool,
+        base_url: &str,
+        analytics_id: Option<&str>,
+    ) -> Result<Self, url::ParseError> {
```
If EC should not be part of the rust sdk.
we can change it to:
```
pub fn new_virtual_widget(url: String, id: String, init_on_load: boolean) -> Self {
    Self { id, init_on_load: false, raw_url }
}
```

toger5 commented on crates/matrix-sdk/src/widget/widget_settings.rs:
```diff
@@ -0,0 +1,186 @@
+
+        // Custom element call url parameters:
+        raw_url.push_str(&format!("&parentUrl={}", encode(&parent_url)));
+        if embed {
+            raw_url.push_str("&embed=")
+        }
```
This has now changed to: `confineToRoom`, `appPrompt=false`

toger5 replied:
```suggestion
        if confined_to_room {
            raw_url.push_str("&confineToRoom=")
        }
        if !add_prompt {
            raw_url.push_str("&addPrompt=false")
        }
```

toger5 commented on crates/matrix-sdk/src/widget/widget_settings.rs:
```diff
@@ -0,0 +1,186 @@
+    ///   its own postmessage messages. The matrix-widget-api (js) ignores those
+    ///   so this works but it might break custom implementations. So always
+    ///   keep this in mind.
+    pub fn new_virtual_element_call_widget(
+        element_call_path: &str,
+        id: String,
+        embed: bool,
+        hide_header: bool,
+        preload: bool,
+        font_scale: f64,
+        analytics_id: Option<&str>,
+        parent_url: &str,
+    ) -> Self {
```
maybe defaults could be added to those params

toger5 commented on crates/matrix-sdk/src/widget/widget_settings.rs:
```diff
@@ -0,0 +1,186 @@
+    // Something like: room.get_widgets() -> Vec<WidgetStateEvent>
+}
+
+/// The set of settings and propterties for the widget based on the client
```
```suggestion
/// The set of settings and properties for the widget based on the client
```

Hywan commented on bindings/matrix-sdk-ffi/src/widget.rs:
```diff
@@ -26,23 +26,175 @@ impl From<Widget> for matrix_sdk::widget::Widget {
+    )
+    .await
+    .map(|url| url.to_string())?)
+}
+/// `WidgetSettings` are usually created from a state event.
```
```suggestion
}

/// `WidgetSettings` are usually created from a state event.
```

Hywan commented on bindings/matrix-sdk-ffi/src/widget.rs:
```diff
@@ -26,23 +26,175 @@ impl From<Widget> for matrix_sdk::widget::Widget {
+    .map(|url| url.to_string())?)
+}
+/// `WidgetSettings` are usually created from a state event.
+/// (currently unimplemented)
```
It seems implemented, no?

toger5 replied:
This implementation generates `WidgetSettings` from fn params. Most widgets will be part of the room state.
What is not yet implemented is will be `room.get_widgets() -> Vec<WidgetSettings>` (where `room: Room`.)

Hywan commented on bindings/matrix-sdk-ffi/src/widget.rs:
```diff
@@ -26,23 +26,175 @@ impl From<Widget> for matrix_sdk::widget::Widget {
+        let ClientProperties { client_id, language_tag, theme } = value;
+        Self::new(&client_id, language_tag, theme)
+    }
+}
 /// Communication "pipes" with a widget.
```
```suggestion
}

/// Communication "pipes" with a widget.
```

Hywan commented on crates/matrix-sdk/src/widget/widget_settings.rs:
```diff
@@ -0,0 +1,217 @@
+    pub static CLIENT_ID: &str = "$org.matrix.msc2873.client_id";
+    pub static DEVICE_ID: &str = "$org.matrix.msc2873.matrix_device_id";
+    pub static BASE_URL: &str = "$org.matrix.msc4039.matrix_base_url";
+}
+/// Settings of the widget.
```
```suggestion
}

/// Settings of the widget.
```

Hywan commented on crates/matrix-sdk/src/widget/widget_settings.rs:
```diff
@@ -0,0 +1,217 @@
+    ///
+    /// * `room` - A matrix room which is used to query the logged in username
+    /// * `props` - Properties from the client that can be used by a widget to
+    ///   adapt to the client. e.g. language, font-scale...
+
+    pub async fn generate_url(
```
```suggestion
    ///   adapt to the client. e.g. language, font-scale...
    pub async fn generate_url(
```

Hywan commented on crates/matrix-sdk/src/widget/widget_settings.rs:
```diff
@@ -0,0 +1,217 @@
+}
+
+impl WidgetSettings {
+    /// Create the actual Url that can be used to setup the WebView or IFrame
```
```suggestion
    /// Create the actual [`Url`] that can be used to setup the WebView or IFrame
```

Hywan commented on crates/matrix-sdk/src/widget/widget_settings.rs:
```diff
@@ -0,0 +1,217 @@
+            .account()
+            .get_profile()
+            .await
+            .unwrap_or(get_profile::v3::Response::new(None, None));
```
```suggestion
            .unwrap_or_default();
```

Hywan commented on crates/matrix-sdk/src/widget/widget_settings.rs:
```diff
@@ -0,0 +1,217 @@
+                .replace(url_props::WIDGET_ID, &self.id)
+                .replace(url_props::AVATAR_URL, &avatar_url)
+                .replace(url_props::DEVICE_ID, &device_id)
+                .replace(url_props::DISPLAY_NAME, &profile.displayname.unwrap_or("".to_owned()))
```
```suggestion
                .replace(url_props::DISPLAY_NAME, &profile.displayname.unwrap_or_default())
```

Hywan commented on crates/matrix-sdk/src/widget/widget_settings.rs:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    /// `WidgetSettings` are usually created from a state event.
+    /// (currently unimplemented)
```
It seems implemented to me :-)?

toger5 replied:
This implementation creates a widget without using matrix room state events.
It is using a list of fn parameters to generate a WidgetSettings object.
It should work like this `Room.getWidgets() -> Vec<WidgetSettings>`

Hywan commented on crates/matrix-sdk/src/widget/widget_settings.rs:
```diff
@@ -0,0 +1,217 @@
+            raw_url.push_str("&embed=")
+        }
+        if hide_header.unwrap_or(true) {
+            raw_url.push_str("&hideHeader=")
```
```suggestion
            raw_url.push_str("&hideHeader=");
```

Hywan commented on crates/matrix-sdk/src/widget/widget_settings.rs:
```diff
@@ -0,0 +1,217 @@
+    pub language: LanguageTag,
+    /// A string describing the theme (dark, light) or org.example.dark.
+    pub theme: String,
+}
+impl ClientProperties {
```
```suggestion
}

impl ClientProperties {
```

toger5 commented on crates/matrix-sdk/src/widget/widget_settings.rs:
```diff
@@ -0,0 +1,217 @@
+    /// * `language` the language that is used in the client. (default: `en-US`)
+    /// * `theme` the theme (dark, light) or org.example.dark. (default:
+    ///   `light`)
+    pub fn new(client_id: &str, language: Option<String>, theme: Option<String>) -> Self {
```
I did not change this afterall. Reason beeing is that this would require the bindings create to also import LanguageTag. And for the ffi we ask for Strings anyways right?

jplatte replied:
It's okay for the FFI API to use less strict types, but that doesn't mean we should use less strict types in the main crates.

toger5 commented:
@jplatte @Hywan 
I refactored this pr and added tests.
The biggest changes are: 
 - I use serde for the query generation (this removes all the slightly awkward string manipulation code. I still dont like the replace chain: https://github.com/matrix-org/matrix-rust-sdk/pull/2487/files#diff-cdd083724293c3e5c29f1a0d549ca6400d2fa6657283b103102d7e1a73c68e44R35-R51, Is there an easy solution?)
 - I moved all ElementCall related code into its own module.
 - Added proper errors. (Created a new uniffi error type for the errors we might encounter)
 - Tests.

jplatte commented on crates/matrix-sdk/Cargo.toml:
```diff
@@ -92,11 +92,13 @@ ruma = { workspace = true, features = ["rand", "unstable-msc2448", "unstable-msc
 serde = { workspace = true }
 serde_html_form = { workspace = true }
 serde_json = { workspace = true }
+serde_urlencoded = "0.7.1"
```
We already depend on `serde_html_form` which is a fork of `serde_urlencoded`, can you please use that one instead?

jplatte commented on crates/matrix-sdk/src/widget/permissions.rs:
```diff
@@ -23,4 +23,8 @@ pub struct Permissions {
     pub read: Vec<EventFilter>,
     /// Types of the messages that a widget wants to be able to send.
     pub send: Vec<EventFilter>,
+    /// If this is set to true the client should not give the option to pop the
+    /// widget into its own window. (The widget will set this to true if it
+    /// relies on the widget-api.)
```
```suggestion
    /// If this is set to true the client should not give the option to pop the
    /// widget into its own window.
    ///
    /// The widget will set this to `true` if it relies on the widget API.
```

Though honestly I have no clue what the second sentence means, maybe some rewording would be helpful as well.

jplatte commented on bindings/matrix-sdk-ffi/src/widget.rs:
```diff
@@ -193,3 +328,50 @@ impl matrix_sdk::widget::PermissionsProvider for PermissionsProviderWrap {
+    SetHostOnCannotBeABaseUrl,
+    #[error("URLs more than 4 GB are not supported")]
+    Overflow,
+    #[error("unknown parse error")]
```
```suggestion
    #[error("unknown URL parsing error")]
```

jplatte commented on crates/matrix-sdk/src/widget/widget_settings/element_call.rs:
```diff
@@ -0,0 +1,123 @@
+}
+
+impl WidgetSettings {
+    /// `WidgetSettings` are usually created from a state event.
+    /// (currently unimplemented)
+    /// But in some cases the client wants to create custom `WidgetSettings`
```
```suggestion
    /// `WidgetSettings` are usually created from a state event
    /// (currently unimplemented).
    ///
    /// In some cases the client wants to create custom `WidgetSettings`
```

jplatte commented on bindings/matrix-sdk-ffi/src/widget.rs:
```diff
@@ -48,25 +48,154 @@ impl WidgetDriver {
+        let (id, init_after_content_load, raw_url) = (
+            value.id().to_owned(),
+            value.init_after_content_load(),
+            value.raw_url().as_str().to_string(),
```
```suggestion
            value.raw_url().to_string(),
```

jplatte commented on crates/matrix-sdk/src/widget/widget_settings/mod.rs:
```diff
@@ -0,0 +1,351 @@
+    }
+    pub fn replace_properties(url: &mut Url, props: QueryProperties) {
+        *url = Url::parse(
+            &url.as_str()
+                .replace(WIDGET_ID, &encode(&props.widget_id))
+                .replace(AVATAR_URL, &encode(&props.avatar_url))
+                .replace(DEVICE_ID, &encode(&props.device_id))
+                .replace(DISPLAY_NAME, &encode(&props.display_name))
+                .replace(HOMESERVER_URL, &encode(&props.homeserver_url))
+                .replace(USER_ID, &encode(&props.user_id))
+                .replace(ROOM_ID, &encode(&props.room_id))
+                .replace(LANGUAGE, &encode(&props.language))
+                .replace(CLIENT_THEME, &encode(&props.client_theme))
+                .replace(CLIENT_ID, &encode(&props.client_id)),
```
Yeah it's pretty unlikely to happen, same for one of the placeholders expanding to another placeholder that could cause trouble now, I just think it would be cleaner to work iteratively and look for the end of the placeholder name before finding a replacement.

toger5 replied:
Building up the string in a for loop is definitly more performant. So I started with that.
What do you think about this usecase: `some/path/Captain_$matrix_display_names_place` where they add a plural s in the url template?

toger5 commented on crates/matrix-sdk/src/widget/widget_settings/mod.rs:
```diff
@@ -0,0 +1,351 @@
+        user_id: String,
+        room_id: String,
+        device_id: String,
+        homeserver_url: String,
```
I changed it to strong types in the method. For the Query properties I left them strings. The `QueryProperties` are only used for actually replacing the palceholders with the parameters so it really only cares about strings.

jplatte commented on bindings/matrix-sdk-ffi/src/widget.rs:
```diff
@@ -48,25 +49,165 @@ impl WidgetDriver {
+        let WidgetSettings { id, init_after_content_load, raw_url } = value;
+        Ok(matrix_sdk::widget::WidgetSettings::new(id, init_after_content_load, &raw_url)?)
+    }
+}
+impl From<matrix_sdk::widget::WidgetSettings> for WidgetSettings {
```
```suggestion
}

impl From<matrix_sdk::widget::WidgetSettings> for WidgetSettings {
```

jplatte commented on crates/matrix-sdk/src/widget/widget_settings/mod.rs:
```diff
@@ -0,0 +1,393 @@
+
+mod element_call;
+
+mod url_params {
```
Can you move this module to its own file please? Also, please move tests for `replace_properties` into the module then (in an inline `mod tests` like in this file).

jplatte commented on crates/matrix-sdk/src/widget/widget_settings/element_call.rs:
```diff
@@ -0,0 +1,125 @@
+    analytics_id: Option<String>,
+    font_scale: Option<f64>,
+    fonts: Option<String>,
+}
+/// Properties to create a new virtual Element Call widget.
```
```suggestion
}

/// Properties to create a new virtual Element Call widget.
```

jplatte commented on crates/matrix-sdk/src/widget/permissions.rs:
```diff
@@ -23,4 +23,11 @@ pub struct Permissions {
     pub read: Vec<EventFilter>,
     /// Types of the messages that a widget wants to be able to send.
     pub send: Vec<EventFilter>,
+    /// If this permission is requested by the widget the client should not
+    /// provide the option to pop the widget into its own window.
+    ///
+    /// If the widget does not work without the widget api (e.g. it relies on
+    /// asking for matrix room messages) it will(/has to) request this
+    /// permission.
```
Still no clue how popping into its own window is related to client permissions, but this doc comment is definitely better than before. Can you copy it over to the FFI crate?

toger5 replied:
I will add some background to this here. Maybe you have an idea what is missing in the doc to make it more clear.
A client (only EW atm) can provide a button to pop out the IFrame into its own window so a widget can be used on a seperate screen for instance. This breaks the widget api communication over postmessage. Some widgets are not usable then (e.g. EC). For a widget to communicate that it wont work anymore if the client decides to pop it out into its own window a capability was introduced. (Its not the most straight forward capability but it is something the widget need from the client to be fully operational. (in this case its inverse. its something the client is NOT allowed to do to make the widget operational). If the widget "requests the capability" (tells the client it will only work if it can use postmessages/ it requires the client) it asks the client for the`io.element.requires_client` capability. Form this point on the client should hide/disable all ux flows that would allow to pop out the widget into its own window. EW just hides the according button.

jplatte replied:
By "its own window", do you mean opening the widget in the system's default browser?

toger5 replied:
yes exacty

jplatte commented on crates/matrix-sdk/src/widget/widget_settings/mod.rs:
```diff
@@ -0,0 +1,393 @@
+    language: LanguageTag,
+    /// A string describing the theme (dark, light) or org.example.dark.
+    theme: String,
+}
+impl ClientProperties {
```
```suggestion
}

impl ClientProperties {
```

jplatte commented on crates/matrix-sdk/src/widget/widget_settings/element_call.rs:
```diff
@@ -0,0 +1,125 @@
+    pub fonts: Option<Vec<String>>,
+    /// Can be used to pass a PostHog id to element call.
+    pub analytics_id: Option<String>,
+}
+impl WidgetSettings {
```
```suggestion
}

impl WidgetSettings {
```

jplatte commented on crates/matrix-sdk/src/widget/permissions.rs:
```diff
@@ -23,4 +23,11 @@ pub struct Permissions {
     pub read: Vec<EventFilter>,
     /// Types of the messages that a widget wants to be able to send.
     pub send: Vec<EventFilter>,
+    /// If this permission is requested by the widget the client should not
+    /// provide the option to pop the widget into its own window.
+    ///
+    /// If the widget does not work without the widget api (e.g. it relies on
+    /// asking for matrix room messages) it will(/has to) request this
+    /// permission.
```
How about this then?

```suggestion
    /// If this permission is requested by the widget, it can not operate
    /// separately from the matrix client.
    ///
    /// This means clients should not offer to open the widget in a separate
    /// browser.
```

toger5 replied:
Yes this is good. I changed `browser` to make it more clear what is meant by this but other then that used your proposal:
```
    /// If this permission is requested by the widget, it can not operate
    /// separately from the matrix client.
    ///
    /// This means clients should not offer to open the widget in a separate
    /// browser/tab/webview that is not connected to the postmessage widget-api.
    ```
